parsing/pcoq.mli

```diff
@@ -41,6 +41,16 @@ end
    - static rules explicitly defined in files g_*.ml4
    - static rules macro-generated by ARGUMENT EXTEND, TACTIC EXTEND and
      VERNAC EXTEND (see e.g. file extratactics.ml4)
+
+   Note that parsing a Coq document is in essence stateful: the parser
+   needs to recognize commands that start proofs and use a different
+   parsing entry point for them.
+
+   We thus provide two different interfaces: the "raw" parsing
+   interface, in the style of camlp5, which provides more flexibility,
+   and a more specialize "parse_vernac" one, which will indeed adjust
+   the state as needed.
+
```

Comment:
I am not sure I fully share the wording here. In a sense all that `parse_vernac` is doing is to choose a different parsing entry point; so indeed maybe `parse_vernac` should become pure and take an extra parameter "proof_mode" such that clients do indeed have to specify the parsing mode.

Long term, indeed, when inside a proof we should never ever have to support vernacs, but tactics [which could piggyback on vernac if wanted].

Reply:
> so indeed maybe `parse_vernac` should become pure and take an extra parameter "proof_mode" such that clients do indeed have to specify the parsing mode.

In fact, this is what we have done, see `Vernacstate.Parser.parse`. It seems to me that the comment is now obsolete.

Did you want another API?

Reply:
I am not sure the "is in essence stateful" is something that I agree with. The parser doesn't need to recognize proofs, it is the clients of the parser that need to do so.

Indeed as I mentioned below I would prefer if the state of `pvernac` is eliminated, and the `Vernacstate.Parser` module would become redundant.

Reply:
To say in different words, for what `Vernacstate.Parser` does, it think it is better to reify the whole imperative state to a parameter.

Reply:
I looked a bit more closely, and I think the current mess is more or less due (as you mention above) to the injection of vernac in tactics, which gives `tactic_entry : Vernacexpr.vernac_expr Entry.t`.

I'm afraid this prevents a clean encapsulation of proof modes at the `Proof_global` level (it has to be done in vernac). I will try to improve the current API for registering and passing proof modes, though.

Reply:
I'm not sure what I should do about the use of `main_entry` in `g_toplevel`. It seems this one can't really use the functional interface, can it?

Reply:
Oh indeed that one is a bit problematic :( I think parse_toplevel could indeed call the functional interface but that'll require a bit of massaging on the entry above.

Reply:
Meanwhile, I can expose the non-functional one (i.e. the real grammar entry) in `Vernac_`, and migrate all other clients to the functional one.